
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Workshop teaching researchers core skills for conducting open science.">
      
      
        <meta name="author" content="CyVerse Science Team">
      
      
        <link rel="canonical" href="https://foss.cyverse.org/nextflow/">
      
      
      
      
      <link rel="icon" href="../assets/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.42">
    
    
      
        <title>Nextflow - CyVerse Foundational Open Science Skills 2024</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.0253249f.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRegular:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Regular"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","UA-72224033-6"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","UA-72224033-6",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=UA-72224033-6",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#automation-workflow-management-with-nextflow-and-snakemake" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="CyVerse Foundational Open Science Skills 2024" class="md-header__button md-logo" aria-label="CyVerse Foundational Open Science Skills 2024" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m20 22-3.86-1.55c.7-1.53 1.2-3.11 1.51-4.72zM7.86 20.45 4 22l2.35-6.27c.31 1.61.81 3.19 1.51 4.72M12 2s5 2 5 10c0 3.1-.75 5.75-1.67 7.83A2 2 0 0 1 13.5 21h-3a2 2 0 0 1-1.83-1.17C7.76 17.75 7 15.1 7 12c0-8 5-10 5-10m0 10c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CyVerse Foundational Open Science Skills 2024
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Nextflow
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/CyVerse-learning-materials/foss" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    CyVerse-learning-materials/foss
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href=".." class="md-tabs__link">
          
  
    
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../01_intro_open_sci/" class="md-tabs__link">
          
  
    
  
  Lessons

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../final_project/overview/" class="md-tabs__link">
          
  
    
  
  Capstone Project

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="CyVerse Foundational Open Science Skills 2024" class="md-nav__button md-logo" aria-label="CyVerse Foundational Open Science Skills 2024" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m20 22-3.86-1.55c.7-1.53 1.2-3.11 1.51-4.72zM7.86 20.45 4 22l2.35-6.27c.31 1.61.81 3.19 1.51 4.72M12 2s5 2 5 10c0 3.1-.75 5.75-1.67 7.83A2 2 0 0 1 13.5 21h-3a2 2 0 0 1-1.83-1.17C7.76 17.75 7 15.1 7 12c0-8 5-10 5-10m0 10c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2"/></svg>

    </a>
    CyVerse Foundational Open Science Skills 2024
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/CyVerse-learning-materials/foss" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    CyVerse-learning-materials/foss
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href=".." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Home
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_1" id="__nav_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Before FOSS Starts
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Schedule
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../code_of_conduct/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Code of Conduct
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../glossary/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Glossary & Acronyms
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Lessons
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Lessons
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_intro_open_sci/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    1. Open Science
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../02_managing_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    2. Data Management
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../03_documentation_communication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    3. Documentation & Communication
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../04_talk_to_computer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    4. How to Talk to Computers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../05_version_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    5. Version Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../06_reproducibility_I/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    6. Reproducibility I: Software Environments
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../07_reproducibility_II/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    7. Reproducibility II: Run Containers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../08_reproducibility_III/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    8. Reproducibility III: Build Containers
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../09_remote_computing_cyverse/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    9. Remote Computing: Cyverse
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../10_reproducibility_IV/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    10. Remote Computing: HPC
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Capstone Project
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Capstone Project
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../final_project/overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  

<nav class="md-nav md-nav--secondary" aria-label="On this page">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      On this page
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#automation-workflow-management-with-nextflow-and-snakemake" class="md-nav__link">
    <span class="md-ellipsis">
      Automation: Workflow Management with Nextflow and Snakemake
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reproducibility-tutorial" class="md-nav__link">
    <span class="md-ellipsis">
      Reproducibility Tutorial
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Reproducibility Tutorial">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prerequisites" class="md-nav__link">
    <span class="md-ellipsis">
      Prerequisites
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#package-management-with-conda" class="md-nav__link">
    <span class="md-ellipsis">
      Package management with Conda
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#package-management-with-pip" class="md-nav__link">
    <span class="md-ellipsis">
      Package management with Pip
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#workflow-tutorial-using-nextflow" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow Tutorial using Nextflow
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Workflow Tutorial using Nextflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#script-1-indexing-transcriptome" class="md-nav__link">
    <span class="md-ellipsis">
      Script 1: Indexing transcriptome
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#script-2-collecting-pairs-and-performing-quantification" class="md-nav__link">
    <span class="md-ellipsis">
      Script 2: collecting pairs and performing quantification
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#script-3-qc-and-report" class="md-nav__link">
    <span class="md-ellipsis">
      Script 3: QC and report
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#essential-exercise-documenting-your-work" class="md-nav__link">
    <span class="md-ellipsis">
      Essential Exercise: Documenting your Work
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Essential Exercise: Documenting your Work">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#github-repository-setup-and-documentation" class="md-nav__link">
    <span class="md-ellipsis">
      GitHub repository setup and documentation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Nextflow</h1>

<h2 id="automation-workflow-management-with-nextflow-and-snakemake">Automation: Workflow Management with Nextflow and Snakemake<a class="headerlink" href="#automation-workflow-management-with-nextflow-and-snakemake" title="Permanent link">&para;</a></h2>
<p>Workflow management is key to automation: it helps with bridging the gap between software, creating useful pipelines that help extracting information from data. Two examples of workflow management system popular (but not restricted to!) within the field of Bioinformatics are <a href="https://www.nextflow.io/docs/latest/index.html">Nextflow</a> and <a href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a>. </p>
<p>Both workflow managers are powerful tools that tackle reproducibility and scalability, however the two have some key differences:</p>
<table>
<thead>
<tr>
<th>Feature</th>
<th>Snakemake</th>
<th>Nextflow</th>
</tr>
</thead>
<tbody>
<tr>
<td>Language style</td>
<td>Python-based DSL*</td>
<td>Groovy-based DSL*</td>
</tr>
<tr>
<td>Execution Model</td>
<td>Rule-based</td>
<td>Dataflow-driven</td>
</tr>
<tr>
<td>Parallelization and Resource Management</td>
<td>Implicit parallelization, limited resource management</td>
<td>Explicit parallelization, more flexible resource management</td>
</tr>
<tr>
<td>Integration</td>
<td>seamless with Python</td>
<td>seamless with Java**</td>
</tr>
</tbody>
</table>
<p>* = Domain Specific Langauge <br>
** = No, it does not mean you need to learn Java</p>
<p>Although there are differences between the two workflow managers, there are a number of features which both share:</p>
<ul>
<li>Available on all OS</li>
<li>Well integrated with Conda</li>
<li>Support the use of Containers</li>
<li>Deployable on HPC systems</li>
<li>Integrated error reporting</li>
<li>Well established communities***</li>
</ul>
<p>*** = It should be noted that Nextflow has a <a href="https://nf-co.re/">community driven repository, nf-core</a>, where people can upload their Nextflow pipelines for others to use. Power to the people!</p>
<p>Both are incredibly powerful when it comes to creating pipelines, ultimately the choice on which to use depends on <strong>you</strong> (and the goal of your project). </p>
<p>Here are a couple of script examples for each workflow manager. Although different, the examples try to </p>
<ol>
<li>Download a genome</li>
<li>Align reads using BWA (<a href="https://bio-bwa.sourceforge.net/">Burrows-Wheeler Aligner</a> tool)</li>
<li>Gather aignment statistics using <a href="https://www.htslib.org/">Samtools</a> (Sequence Alignment/Map)</li>
</ol>
<p><strong>Snakemake:</strong></p>
<div class="highlight"><pre><span></span><code># Snakefile

# Define rule for downloading reference genome
rule download_reference_genome:
    output:
        &quot;reference_genome.fa&quot;
    shell:
        &quot;wget http://example.com/reference_genome.fa&quot;

# Define rule for aligning reads to the reference genome using BWA
rule align_reads:
    input:
        &quot;reference_genome.fa&quot;,
        fastq=&quot;data/{sample}.fastq&quot;
    output:
        &quot;aligned/{sample}.bam&quot;
    shell:
        &quot;bwa mem {input.reference_genome} {input.fastq} | samtools view -b - &gt; {output}&quot;

# Define rule to aggregate alignment statistics
rule aggregate_stats:
    input:
        expand(&quot;aligned/{sample}.bam&quot;, sample=config[&#39;samples&#39;])
    output:
        &quot;alignment_stats.txt&quot;
    shell:
        &quot;samtools flagstat {input} &gt; {output}&quot;

# Define samples
configfile: &quot;config.yaml&quot;
</code></pre></div>
<p>Notice how in the Snakefile above, there are a set number of rules that define input, output and commands.</p>
<p><strong>Nextflow:</strong></p>
<div class="highlight"><pre><span></span><code>// main.nf

params.samples = [&#39;sample1&#39;, &#39;sample2&#39;, &#39;sample3&#39;]

// Define process to download reference genome
process download_reference_genome {
    output:
    file(&#39;reference_genome.fa&#39;)

    script:
    &quot;&quot;&quot;
    wget http://example.com/reference_genome.fa -O reference_genome.fa
    &quot;&quot;&quot;
}

// Define process to align reads using BWA
process align_reads {
    input:
    file(reference_genome), file(fastq) from fastqs

    output:
    file(&quot;aligned/${sample}.bam&quot;)

    script:
    &quot;&quot;&quot;
    bwa mem ${reference_genome} ${fastq} | samtools view -b - &gt; aligned/${sample}.bam
    &quot;&quot;&quot;
}

// Define process to aggregate alignment statistics
process aggregate_stats {
    input:
    set file(aligned_bam) from aligned_bams

    output:
    file(&quot;alignment_stats.txt&quot;)

    script:
    &quot;&quot;&quot;
    samtools flagstat ${aligned_bam} &gt; alignment_stats.txt
    &quot;&quot;&quot;
}

// Define workflow execution
workflow {
    // Define input data
    Channel.fromPath(&quot;data/*.fastq&quot;).set { fastqs }

    // Execute processes in parallel
    download_reference_genome()
    align_reads(fastqs)
    aggregate_stats()
}
</code></pre></div>
<p>The Nextflow file "forces" you to know what you are expecting prior to execution by first defining what you want processes you need before defining the workflow script. </p>
<p>Although not shown here, Nextflow uses "Channels" to help with workflow execution. A channel can have one or more defined processes, but it can "wait": these processes wait until the input is ready before executing, thus creating the possibilty of an asynchronous processing pipeline (and parallel processing!). This is also what makes Nextflow able to manage resources better than Snakemake (the processes in "inactive" channels do not use resources until they are needed). </p>
<h2 id="reproducibility-tutorial">Reproducibility Tutorial<a class="headerlink" href="#reproducibility-tutorial" title="Permanent link">&para;</a></h2>
<p>This section is going to cover a short tutorial on reproducibility using software, tools and practices discussed today and throughout FOSS.</p>
<div class="admonition note">
<p class="admonition-title">OS of choice</p>
<p>This tutorial will be performed using the <a href="https://de.cyverse.org/apps/de/5f2f1824-57b3-11ec-8180-008cfa5ae621">CyVerse CLI (Command Line Interface)</a>. However, if you'd like to use your own computer feel free to! If you're on Mac or Linux, open your terminal; If you're on Windows, use the Windows Subsystem for Linux (WSL)</p>
<details class="tip">
<summary>How to Scroll in Cyverse(Tmux) Cloud Shell</summary>
<p>If you're using the Cyverse Cloud Shell, you can scroll up and down by pressing <code>Ctrl + b</code> and then <code>[</code> to enter scroll mode. You can then use the arrow keys to scroll up and down. Press <code>q</code> to exit scroll mode.</p>
</details>
</div>
<div class="admonition success">
<p class="admonition-title">Tutorial Goals</p>
<ul>
<li>Run a small workflow using NextFlow</li>
<li>Understand best practices for reproducing a workflow</li>
<li>Apply FOSS procedures in order to enable easiness of reproducibility</li>
</ul>
</div>
<h3 id="prerequisites">Prerequisites<a class="headerlink" href="#prerequisites" title="Permanent link">&para;</a></h3>
<p>What you'll be using:</p>
<ul>
<li><a href="https://github.com/">GitHub</a> (already installed)</li>
<li><a href="https://docs.conda.io/en/latest/">Conda</a></li>
<li><a href="https://mamba.readthedocs.io/en/latest/installation.html">Mamba</a> (optional, recommended)</li>
</ul>
<p>Installable through Conda/Mamba:</p>
<ul>
<li><a href="https://www.nextflow.io/docs/latest/index.html">Nextflow</a></li>
<li><a href="https://salmon.readthedocs.io/en/latest/">Salmon</a></li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></li>
</ul>
<p>Installabe through Pip:</p>
<ul>
<li><a href="https://multiqc.info/docs/">MultiQC</a></li>
</ul>
<details class="question">
<summary>What's a Conda and how do I install it?</summary>
<p><a href="https://docs.conda.io/en/latest/">Conda</a> is a popular tool for installing software. Typically software you want to use requires other software (dependancies) to be installed. Conda can manage all of this for you. Each available Conda package is part of a “recipe” that includes everything you need to run your software. There are different versions of Conda, including some specific for bioinformatics like <a href="https://bioconda.github.io/">Bioconda</a>.</p>
<p>The CyVerse CLI already comes with Conda installed; Please follow these steps in order to install <a href="https://docs.conda.io/en/latest/miniconda.html">MiniConda</a> (the lightweight version of Conda) on your system.</p>
<p>For the appropriate installation package, visit <a href="https://docs.conda.io/en/latest/miniconda.html">https://docs.conda.io/en/latest/miniconda.html</a>. <img alt="⚠" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.0.3/assets/svg/26a0.svg" title=":warning:" /> Note: <strong>If you are using the WSL, install the Linux version!!</strong></p>
<div class="highlight"><pre><span></span><code># Download conda and add right permissions
wget https://repo.anaconda.com/miniconda/Miniconda3-py39_4.12.0-Linux-x86_64.sh     # Modify this to match the OS you&#39;re using.
chmod +x Miniconda3-py39_4.12.0-Linux-x86_64.sh

# install conda silenty (-b) and update (-u) and initial conda run
./Miniconda3-py39_4.12.0-Linux-x86_64.sh -b -u
~/miniconda3/bin/conda init

# Restart bash so that conda is activated
source ~/.bashrc
</code></pre></div>
<p>You'll be able to tell when conda is active when next <code>(base)</code> is present next to the to the shell prompt such as</p>
<div class="highlight"><pre><span></span><code>(base) user@machine
</code></pre></div>
<p>Conda should now be installed and can be used to install other necessary packages! </p>
</details>
<p>When you start a Cyverse Cloud shell, it will start you in the directory: 
<div class="highlight"><pre><span></span><code>/home/user/work 
</code></pre></div>
Let's change to the Data-Store directory, where we will be working for the rest of the tutorial. This is the Cyverse cloud-storage directory where you can put all your data and files. </p>
<p><div class="highlight"><pre><span></span><code>cd home
cd &lt;username&gt;
</code></pre></div>
Create our own environment (select <code>y</code> when prompted).</p>
<div class="highlight"><pre><span></span><code>conda create --name myenv
</code></pre></div>
<p>Activate your new environment with </p>
<div class="highlight"><pre><span></span><code>conda activate myenv
</code></pre></div>
<p>You can see the list of environments you can activate by doing</p>
<div class="highlight"><pre><span></span><code>conda env list
</code></pre></div>
<h3 id="package-management-with-conda">Package management with Conda<a class="headerlink" href="#package-management-with-conda" title="Permanent link">&para;</a></h3>
<p>We are going to to use conda to install Mamba, NextFlow, Salmon and FastQC.</p>
<div class="highlight"><pre><span></span><code># Activate Conda using 
conda activate

# Install Mamba
conda install -c conda-forge mamba
</code></pre></div>
<p>You can either use the anaconda website to search for packager, or use the <a href="https://docs.conda.io/projects/conda/en/latest/commands/search.html">conda search feature</a> (but also, Google is your best friend.)</p>
<div class="admonition note">
<p class="admonition-title">Makes things faster with Mamba</p>
<p>Mamba is a reimplemetation of Conda using the C++ language, allowing for much quicker Conda experience. The tutorial is going to use <strong>Mamba</strong> instead of <strong>Conda</strong>, but you can always replace <code>mamba</code> with <code>conda</code>!</p>
</div>
<div class="admonition info">
<p class="admonition-title">Conda channels</p>
<p>Conda operates through <strong>channels</strong>, specififc repositories where packages are stored. Specific packages sometimes may appear in multiple channels, however it is always helpful to specify a channel with the <code>-c</code> flag.</p>
</div>
<p>Install Nextflow and verify its installation with the following commands: </p>
<div class="highlight"><pre><span></span><code># Install NextFlow
mamba install -c bioconda nextflow=22.10.6     # Press y when prompted with [Y/n]!

# verify the installation
nextflow -version
</code></pre></div>
<p>Now that you know how to install packages with Conda/Mamba, install Salmon and FastQC.</p>
<details class="question">
<summary>Installing Packages</summary>
<p>As an exercise, install Salmon and FastQC using Conda/Mamba.</p>
<details class="tip">
<summary>Need a hand?</summary>
<div class="highlight"><pre><span></span><code>mamba install -c bioconda salmon
mamba install -c bioconda fastqc
</code></pre></div>
<p>Or you can do it with a single line (doable if packages are from the <em>same</em> channel)!
<div class="highlight"><pre><span></span><code>mamba install -c bioconda salmon fastqc
</code></pre></div></p>
</details>
</details>
<p>You can view the installed conda packages by doing </p>
<div class="highlight"><pre><span></span><code>conda list
</code></pre></div>
<p>In order to make your environment reproducible, conda allows you to export your environment.</p>
<div class="highlight"><pre><span></span><code>conda env export &gt; my_conda_env.yml
</code></pre></div>
<h3 id="package-management-with-pip">Package management with Pip<a class="headerlink" href="#package-management-with-pip" title="Permanent link">&para;</a></h3>
<p>Pip works similarly to Conda, as Pip is the package management supported by the Python Software foundation. If you use Python for your work it is likely you have installed packages using Pip.</p>
<p>We only have to install a single package required for this tutorial, MultiQC. To install MultiQC using Pip, do:</p>
<div class="highlight"><pre><span></span><code>pip install multiqc
</code></pre></div>
<p>Similar to Conda, you can export your pip environment by doing</p>
<div class="highlight"><pre><span></span><code>pip3 freeze &gt; my_pip_env.txt
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Why <code>pip3</code>?</p>
<p><code>pip3 freeze &gt; my_pip_env.txt</code> is used to export the pip environment such that it is readable for Python 3. If you want to export an environment for Python 2, you can use <code>pip freeze &gt; my_pip_env.txt</code>.</p>
</div>
<div class="admonition success">
<p class="admonition-title">Conda exports your Pip environment as well</p>
<p>Exporting your environment using Conda (<code>conda env export &gt; my_conda_env.yml</code>) will <strong>ALSO</strong> export your pip environment!</p>
</div>
<h3 id="workflow-tutorial-using-nextflow">Workflow Tutorial using Nextflow<a class="headerlink" href="#workflow-tutorial-using-nextflow" title="Permanent link">&para;</a></h3>
<div class="admonition info">
<p class="admonition-title">...what are we doing?</p>
<p>In this tutorial (now that we have set up the environment, repository and pushed our first commit) we are going to:</p>
<ul>
<li>Index a <a href="https://en.wikipedia.org/wiki/Transcriptome">transcriptome</a> file and quantification of DNA reads (using <a href="https://salmon.readthedocs.io/en/latest/">Salmon</a>).</li>
<li>Perform quality controls (with <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>).</li>
<li>Create a <a href="https://multiqc.info/docs/">MultiQC</a> report.</li>
</ul>
</div>
<div class="admonition warning">
<p class="admonition-title">I Don't know DNA stuff, can I still do this?</p>
<p>Absolutely yes! This tutorial is supposed to introduce you to the process of reproducibility using GitHub repositories, package managers and workflow managers! You do not need to understand what each file is as this is meant to show how to make your science reproducible. Focus on understanding the process and theory behind the tutorial rather than the files themselves <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M464 256a208 208 0 1 0-416 0 208 208 0 1 0 416 0M0 256a256 256 0 1 1 512 0 256 256 0 1 1-512 0m177.6 62.1c15.2 16.4 41.2 33.9 78.4 33.9s63.2-17.5 78.4-33.9c9-9.7 24.2-10.4 33.9-1.4s10.4 24.2 1.4 33.9c-22 23.8-60 49.4-113.6 49.4s-91.7-25.5-113.6-49.4c-9-9.7-8.4-24.9 1.4-33.9s24.9-8.4 33.9 1.4zm40-89.3-.2-.2c-.2-.2-.4-.5-.7-.9-.6-.8-1.6-2-2.8-3.4-2.5-2.8-6-6.6-10.2-10.3-8.8-7.8-18.8-14-27.7-14s-18.9 6.2-27.7 14c-4.2 3.7-7.7 7.5-10.2 10.3-1.2 1.4-2.2 2.6-2.8 3.4-.3.4-.6.7-.7.9l-.2.2c-2.1 2.8-5.7 3.9-8.9 2.8s-5.5-4.1-5.5-7.6c0-17.9 6.7-35.6 16.6-48.8 9.8-13 23.9-23.2 39.4-23.2s29.6 10.2 39.4 23.2c9.9 13.2 16.6 30.9 16.6 48.8 0 3.4-2.2 6.5-5.5 7.6s-6.9 0-8.9-2.8m160 0-.2-.2c-.2-.2-.4-.5-.7-.9-.6-.8-1.6-2-2.8-3.4-2.5-2.8-6-6.6-10.2-10.3-8.8-7.8-18.8-14-27.7-14s-18.9 6.2-27.7 14c-4.2 3.7-7.7 7.5-10.2 10.3-1.2 1.4-2.2 2.6-2.8 3.4-.3.4-.6.7-.7.9l-.2.2c-2.1 2.8-5.7 3.9-8.9 2.8s-5.5-4.1-5.5-7.6c0-17.9 6.7-35.6 16.6-48.8 9.8-13 23.9-23.2 39.4-23.2s29.6 10.2 39.4 23.2c9.9 13.2 16.6 30.9 16.6 48.8 0 3.4-2.2 6.5-5.5 7.6s-6.9 0-8.9-2.8"/></svg></span>.</p>
</div>
<p>Nextflow is a workflow manager, similar to <a href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a>. For this tutorial, we decided to use Nextflow as it is easier to learn, more intuitive and user friendly than Snakemake.</p>
<p>Download the required files using <code>wget</code> and <code>tar</code> to decompress them</p>
<div class="highlight"><pre><span></span><code>wget -O nf_foss_tut.tar.gz https://github.com/CyVerse-learning-materials/foss/blob/mkdocs/docs/assets/tutorials/nf_foss_tut.tar.gz?raw=true
tar -xvf nf_foss_tut.tar.gz
</code></pre></div>
<p>We can now look at the decompressed directory structure by using <code>tree nf_foss_tut</code> (if you don not have <code>tree</code> installed, you can install it with <code>sudo apt-get tree</code> or <code>mamba install -c conda-forge tree</code>).</p>
<div class="highlight"><pre><span></span><code>.
├── nf_foss_tut
│   ├── data
│   │   └── ggal
│   │       ├── gut_1.fq
│   │       ├── gut_2.fq
│   │       ├── liver_1.fq
│   │       ├── liver_2.fq
│   │       ├── lung_1.fq
│   │       ├── lung_2.fq
│   │       └── transcriptome.fa
│   ├── example_script.nf
│   ├── script1.nf
│   ├── script2.nf
│   └── script3.nf

2 directories, 11 files
</code></pre></div>
<div class="admonition info">
<p class="admonition-title">Files information</p>
<ul>
<li>Scripts 1 through 3 (<code>script&lt;number&gt;.nf</code>) and <code>example_script.nf</code> are the NextFlow files</li>
<li><code>&lt;file&gt;.fq</code> are fastq files, containing DNA sequences and quality scores</li>
<li><code>transcriptome.fa</code> is all of the RNA data from the organism (<em>G.gallus</em>)</li>
</ul>
</div>
<p>Let's look at one of the NextFlow files (<code>.nf</code>)</p>
<div class="admonition info">
<p class="admonition-title">Understanding the Nextflow synthax</p>
<p>Nextflow is powerful workflow manager as it can be deployed on HPCs and Clouds. However, it does require a little effort in order to understand its synthax. </p>
<p>The synthax is broken down into:</p>
<ul>
<li>Defining parameters early</li>
<li>Defining Processes to be executed</li>
<li>Defining Channels (blocks that work asynchronously that encapsulate other processes)</li>
</ul>
<p>More complex scripts include <a href="https://www.nextflow.io/docs/latest/operator.html">operators</a> (channel manipulation) and <a href="https://www.nextflow.io/docs/latest/executor.html">executors</a> (to run things on the cloud and HPC); Nextflow can also be used to run and orchestrate <a href="https://www.nextflow.io/docs/latest/container.html">containers</a>.</p>
<p>As a good example, let's look at <code>example_script.nf</code>:
<div class="highlight"><pre><span></span><code>/*                                                                                      
    * pipeline input parameters                                                            
    */                                                                                      
    params.reads = &quot;$baseDir/data/ggal/gut_{1,2}.fq&quot;                                        #
    params.transcriptome = &quot;$baseDir/data/ggal/transcriptome.fa&quot;                            # The parameters are set early in the script
    params.multiqc = &quot;$baseDir/multiqc&quot;                                                     #
    params.outdir = &quot;results&quot;                                                               #

    println &quot;&quot;&quot;\                                                                            #
            R N A S E Q - N F   P I P E L I N E                                             #
            ===================================                                             #
            transcriptome: ${params.transcriptome}                                          # Print statement that will show once the script
            reads        : ${params.reads}                                                  # is executed
            outdir       : ${params.outdir}                                                 #
            &quot;&quot;&quot;                                                                             #
            .stripIndent()                                                                  #

    /* 
     * create a transcriptome file object given then transcriptome string parameter
     */
    transcriptome_file = file(params.transcriptome)                                         # Convert input file to string

    /* 
     * define the `index` process that create a binary index 
     * given the transcriptome file
     */
    process index {                                                                         # First process, named &quot;index&quot;
    conda &quot;bioconda::salmon&quot;                                                                # Defines what package is necessary
                                                                                            #
    input:                                                                                  ## 
    file transcriptome from transcriptome_file                                              ##
                                                                                            ## Defines the input and output of the process
    output:                                                                                 ##
    file &#39;index&#39; into index_ch                                                              ##
                                                                                            #
    script:                                                                                 #
    &quot;&quot;&quot;                                                                                     #
    salmon index --threads $task.cpus -t $transcriptome -i index                            # Command to execute
    &quot;&quot;&quot;                                                                                     #
    }                                                                                       #


    Channel                                                                                 # Channels allows for scripts to work asynchronously, without waiting for received process.
    .fromFilePairs( params.reads )                                                          # .fromFilePairs method creates a channel emitting the file pairs matching a &quot;glob&quot; pattern provided by the user. 
    .ifEmpty { error &quot;Cannot find any reads matching: ${params.reads}&quot;  }                   # .ifEmpty emits a value specified if no input is found.
    .set { read_pairs_ch }                                                                  # .set operator assigns the channel to a variable whose name is specified as a closure parameter.
                                                                                            #                
    process quantification {                                                                # Second process, named &quot;quantification&quot;
    conda &quot;bioconda::salmon&quot;                                                                # Defines what package is necessary
    input:                                                                                  #
    file index from index_ch                                                                ##
    set pair_id, file(reads) from read_pairs_ch                                             ##
                                                                                            ## Defines the input and output of the process
    output:                                                                                 ##
    file(pair_id) into quant_ch                                                             ##
                                                                                            #
    script:                                                                                 #
    &quot;&quot;&quot;                                                                                                     #
    salmon quant --threads $task.cpus --libType=U -i index -1 ${reads[0]} -2 ${reads[1]} -o $pair_id        # Command to execute
    &quot;&quot;&quot;                                                                                                     #
    }        
</code></pre></div></p>
<p>Nextflow has in-depth documentation that can be found <a href="https://www.nextflow.io/docs/latest/">here</a>.</p>
</div>
<p>The 3 scripts' tasks are:</p>
<ul>
<li>Script 1 creates the transcriptome index file, necessary for downstream processes.</li>
<li>Script 2 collects read files by pairs (fastq files come in pairs) and performs quantification.</li>
<li>Script 3 performs quality control and summarizes all findings in a single report.</li>
</ul>
<h4 id="script-1-indexing-transcriptome">Script 1: Indexing transcriptome<a class="headerlink" href="#script-1-indexing-transcriptome" title="Permanent link">&para;</a></h4>
<p>Execute script 1
<div class="highlight"><pre><span></span><code>nextflow run script1.nf
</code></pre></div></p>
<p>The output will be something similar to
<div class="highlight"><pre><span></span><code>N E X T F L O W  ~  version 22.10.6
Launching `script1.nf` [admiring_banach] DSL1 - revision: 66baaf0091
R N A S E Q - N F   P I P E L I N E    
===================================
transcriptome: /home/user/work/folder/nf_foss_tut/data/ggal/transcriptome.fa
reads        : /home/user/work/folder/nf_foss_tut/data/ggal/*_{1,2}.fq
outdir       : results

executor &gt;  local (1)
[f0/0a72bc] process &gt; index [100%] 1 of 1 ✔
</code></pre></div>
This is Nextflow's way of telling you that the process has been executed and completed. You should now have a new folder called <code>work</code>. Execute <code>tree work</code> to see what is inside the folder.
<div class="highlight"><pre><span></span><code>work
└── f0
    └── 0a72bc4d10dba1df2899b0449519e9
        ├── index
        │   ├── duplicate_clusters.tsv
        │   ├── hash.bin
        │   ├── header.json
        │   ├── indexing.log
        │   ├── quasi_index.log
        │   ├── refInfo.json
        │   ├── rsd.bin
        │   ├── sa.bin
        │   ├── txpInfo.bin
        │   └── versionInfo.json
        └── transcriptome.fa -&gt; /home/user/work/folder/nf_foss_tut/data/ggal/transcriptome.fa

3 directories, 11 files
</code></pre></div>
These are new index files from the transcriptome provided.</p>
<h4 id="script-2-collecting-pairs-and-performing-quantification">Script 2: collecting pairs and performing quantification<a class="headerlink" href="#script-2-collecting-pairs-and-performing-quantification" title="Permanent link">&para;</a></h4>
<p>Execute with
<div class="highlight"><pre><span></span><code>nextflow run script2.nf -resume --reads &#39;data/ggal/*_{1,2}.fq&#39;
</code></pre></div></p>
<p>The output should look like
<div class="highlight"><pre><span></span><code>N E X T F L O W  ~  version 22.10.6
Launching `script2.nf` [stupefied_swirles] DSL2 - revision: d3b0d0121c
R N A S E Q - N F   P I P E L I N E    
===================================
transcriptome: /home/user/work/folder/nf_foss_tut/data/ggal/transcriptome.fa
reads        : data/ggal/*_{1,2}.fq
outdir       : results

executor &gt;  local(3)                                                                                
[c1/6ece54] process &gt; index [100%] 1 of 1, cached: 1 ✔                          
[1b/10b8d5] process &gt; quantification (1) [100%] 3 of 3 ✔
</code></pre></div></p>
<h4 id="script-3-qc-and-report">Script 3: QC and report<a class="headerlink" href="#script-3-qc-and-report" title="Permanent link">&para;</a></h4>
<p>Execute with
<div class="highlight"><pre><span></span><code>nextflow run script3.nf -resume --reads &#39;data/ggal/*_{1,2}.fq&#39;
</code></pre></div></p>
<p>The output should look like 
<div class="highlight"><pre><span></span><code>N E X T F L O W  ~  version 22.10.6
Launching `script3.nf` [voluminous_goodall] DSL1 - revision: d118356290
R N A S E Q - N F   P I P E L I N E    
===================================
transcriptome: /home/user/work/folder/nf_foss_tut/data/ggal/transcriptome.fa
reads        : data/ggal/*_{1,2}.fq
outdir       : results

executor &gt;  local (4)
[c1/6ece54] process &gt; index                   [100%] 1 of 1, cached: 1 ✔
[7a/4e9ce4] process &gt; quantification (lung)   [100%] 3 of 3, cached: 3 ✔
[34/d60dbb] process &gt; fastqc (FASTQC on lung) [100%] 3 of 3 ✔
[e9/e7c392] process &gt; multiqc                 [100%] 1 of 1 ✔

Done! Open the following report in your browser --&gt; results/multiqc_report.html
</code></pre></div></p>
<p>As you can notice, the report is an <code>html</code> file that can be opened with a browser. Navigate to this file in the Cyverse Data Store and open it. </p>
<hr />
<h3 id="essential-exercise-documenting-your-work">Essential Exercise: Documenting your Work<a class="headerlink" href="#essential-exercise-documenting-your-work" title="Permanent link">&para;</a></h3>
<p><strong>Document your work.</strong> You should still be in your GitHub folder. Summarize your steps and work on your README file, and push your changes! This will ensure that your work and files are saved and have a valid version that you can come back to in the future if you ever require to.</p>
<div class="admonition warning">
<p class="admonition-title">Prerequisites</p>
<p>You will require the following in case you want to add code locally.</p>
<details class="info">
<summary>Create a GitHub account</summary>
<p>Navigate to the <a href="https://github.com/">GitHub website</a> and click <em>Sign Up</em>, and follow the on screen instructions.</p>
</details>
<p><strong>Installing Git</strong></p>
<p>You can follow the official guidelines here: <a href="https://github.com/git-guides/install-git">https://github.com/git-guides/install-git</a>. Here we recommend how to install Git on your local machine.</p>
<details class="info">
<summary>Windows</summary>
<div class="admonition warning">
<p class="admonition-title">These instructions are for Windows users <strong>NOT</strong> using <a href="https://learn.microsoft.com/en-us/windows/wsl/install">WSL2</a>. If you do have WSL2, follow the <strong>Unix</strong> instructions.</p>
</div>
<ol>
<li>Navigate to the <a href="https://gitforwindows.org/">latest Git for Windows installer</a> and download the latest version.</li>
<li>Once the installer has started, follow the instructions as provided in the Git Setup wizard screen until the installation is complete.</li>
<li>Search and open <strong>Git Bash</strong>. From here, you should be able to run Git commands.</li>
</ol>
</details>
<details class="info">
<summary>MacOS</summary>
<ol>
<li>Install <a href="https://brew.sh/">Homebrew</a> (a package manager for MacOS): `/bin/bash -c "$(curl -fsSL <a href="https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh">https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh</a>)"``</li>
<li>Install Git: <code>brew install git</code></li>
</ol>
</details>
<details class="info">
<summary>Unix</summary>
<p>The following command will install git and all related packages on your Unix machine.
<div class="highlight"><pre><span></span><code>$ sudo apt-get install -y git-all
</code></pre></div></p>
</details>
<p>Additionally, you can choose between Generating a Personal Access Token or using SSH keys. This is useful if you want to work locally and push your changes to GitHub. We are going to cover this further in next week's lesson on <a href="../05_version_control/">Version Control</a>.</p>
<details class="info">
<summary>Choice A: Generate a Personal Access Token</summary>
<p>You can follow the official documentation on how to generate Tokens <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens">here</a>. We  discussed how to generate tokens in <a href="https://foss.cyverse.org/00_basics/#adding-code-locally">Week 0</a>. Here's are quick steps you can follow in order to setup your account on your machine using tokens:</p>
<ol>
<li>On your coumputer:<ol>
<li>Clone your repository (<code>git clone &lt;repository&gt;</code>)</li>
<li>Make changes where necessary, and <strong>add</strong> (<code>git add &lt;changed files&gt;</code>), <strong>commit</strong> (<code>git commit -m "&lt;message on changes&gt;"</code>) and <strong>push</strong> your changes (<code>git push origin</code>).</li>
<li>You should be prompted to logging in your GitHub account. Put your email <strong>but not your password</strong>. Instead, open your web browser and follow the steps below:</li>
</ol>
</li>
<li>On GitHub:<ol>
<li>Navigate to your GitHub Settings (You can access your account Settings from the drop down menu where your account icon is, on the top right of the screen)</li>
<li>Scroll to the bottom of the left hand side menu to find <em>Developer settings</em> and open it.</li>
<li>Click <em>Personal access tokens</em> &gt; <em>Tokens (classic)</em></li>
<li>Click <em>Generate new token</em> &gt; <em>Generate new token (classic)</em>. You might need to input your Authentification code if you have enabled 2FA.</li>
<li>Give it a name, and all the scopes you require (tip: select all scopes and <em>No Expiration</em>), then click <em>Generate Token</em>. <strong>Copy the new generated Token</strong></li>
</ol>
</li>
<li>Back on your computer:<ol>
<li>If you have been following the steps above, you should still be in your shell with GitHub still asking for your password.</li>
<li><strong>Paste</strong> your Token here, and you should be logging in. Your changes should then be saved to GitHub.</li>
</ol>
</li>
</ol>
</details>
<details class="info">
<summary>Choice B: Connecting via SSH</summary>
<p>The process of connecting your computer to GitHub using an SSH key is more expedited (and probably less confusing). </p>
<p>As a setup step, see if your computer is already connected to GitHub by doing <code>ssh -T git@github.com</code>. If the response message is <code>git@github.com: Permission denied (publickey).</code> it signifies that your computer is not yet linked with GitHub. To link your computer to github to the following:</p>
<ol>
<li>Generate an SSH key with a level of encryption that you prefer: <code>ssh-keygen -t ed25519 -C &lt;your github email&gt;</code>. This command generates an SSH key with <a href="https://ed25519.cr.yp.to/">ed25519</a> encryption (harder to crack!) and adds your email as "comment" (<code>-C</code>, will help recongizing the user adding the key). A number of additional questions are going to ask you where you'd like to save the key and whether you'd like to add a password for protection; unless you want to save it elsewhere, feel free to use the default options. Upon completion you should see something like this:
<div class="highlight"><pre><span></span><code>Your identification has been saved in /c/Users/&lt;user&gt;/.ssh/id_ed25519
Your public key has been saved in /c/Users/&lt;user&gt;/.ssh/id_ed25519.pub
The key fingerprint is:
SHA256:SMSPIStNyA00KPxuYu94KpZgRAYjgt9g4BA4kFy3g1o &lt;your github email&gt;
The key&#39;s randomart image is:
+--[ED25519 256]--+
|^B== o.          |
|%*=.*.+          |
|+=.E =.+         |
| .=.+.o..        |
|....  . S        |
|.+ o             |
|+ =              |
|.o.o             |
|oo+.             |
+----[SHA256]-----+
</code></pre></div></li>
<li>Upon generating the ssh key, copy it. You can reveal it by doing <code>cat ~/.ssh/id_ed25519.pub</code>.</li>
<li>In GitHub, go to your settings: click your account icon on top right, and from the drop down menu, select <em>Settings</em> and then <em>SSH and GPG keys</em>. Here, click on <em>New SSH Key</em>, where you can then paste the newly geneated key. Add a name reflecting your machine and save changes. </li>
</ol>
<p>Optional: if you want to check if you successfully linked your computer to GitHub, do <code>ssh -t git@github.com</code>. You should receive the following message: `Hi <GitHub username>! You've successfully authenticated, but GitHub does not provide shell access.</p>
</details>
</div>
<details class="question">
<summary>What if my files are too big?</summary>
<p>You can always use a <code>.gitignore</code>, a file that within itself has defined what should be saved in GitHub when pushing a commit, and what shouldn't be saved. An alternative is to move your files outside of the respository that you're pushing ("stashing").</p>
</details>
<h4 id="github-repository-setup-and-documentation">GitHub repository setup and documentation<a class="headerlink" href="#github-repository-setup-and-documentation" title="Permanent link">&para;</a></h4>
<p>Create a repository on GitHub to document your work:</p>
<ul>
<li>On <a href="https://github.com/">GitHub</a>, navigate to your account page and create a new repository (add a README to create structure!)</li>
<li>Clone your repository locally with <code>git clone &lt;repository_url&gt;.git</code> (find the url under the green <strong>Code</strong> button)</li>
<li>Navigate to your cloned repository with <code>cd &lt;repository_name&gt;</code>. You should now be inside your repository.</li>
<li>Move your environemnt files into your repository with <code>mv ../my_conda_env.yml ../my_pip_env.txt .</code>.</li>
<li>Modify your README to reflect the work so far, with meaningful comments (remember that the README is formatted with markdown, a guide to markdown <a href="https://www.markdownguide.org/basic-syntax/">here</a>). A well documented document may look similar to:</li>
</ul>
<div class="highlight"><pre><span></span><code># reproducibility-tutorial

This repository contains information about the reproduciblility tutorial from [FOSS 2023 Spring](https://foss.cyverse.org/06_reproducibility_i/#reproducibility-tutorial).

## Environment Setup

- Download conda and add right permissions
```
wget https://repo.anaconda.com/miniconda/Miniconda3-py39_4.12.0-Linux-x86_64.sh
chmod +x Miniconda3-py39_4.12.0-Linux-x86_64.sh
```
- Install conda silenty (-b), update (-u) and initial start
```
./Miniconda3-py39_4.12.0-Linux-x86_64.sh -b -u
~/miniconda3/bin/conda init
```
-  Restart bash so that conda is activated
```
source ~/.bashrc
```
- Install Mamba
```
conda install -c conda-forge mamba
```
- Use environment files in this repo to recreate tutorial env
```
mamba install -f &lt;my_conda_env.yml&gt;     # Will also install pip packages
```

## Obtaining tutorial files

Tutorial files avaiable [here](https://github.com/CyVerse-learning-materials/foss/blob/mkdocs/docs/assets/tutorials/nf_foss_tut.tar.gz?raw=true). Use `wget` to download appropriate files and decompress files with `tar -xvf`.
```
wget -O nf_foss_tut.tar.gz https://github.com/CyVerse-learning-materials/foss/blob/mkdocs/docs/assets/tutorials/nf_foss_tut.tar.gz?raw=true
tar -xvf nf_foss_tut.tar.gz
```

## Workflow tutorial using Nextflow
Steps of the nextflow tutorial will be added in future commits.
</code></pre></div>
<ul>
<li>Add, commit and push your changes</li>
</ul>
<div class="highlight"><pre><span></span><code>git add .
git commit -m &quot;documenting the tutorial&quot;
git push
</code></pre></div>
<hr />







  
    
  
  


  <aside class="md-source-file">
    
      
  <span class="md-source-file__fact">
    <span class="md-icon" title="Last update">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg>
    </span>
    2024-10-24
  </span>

    
    
    
    
  </aside>


  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2024 CyVerse
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://orcid.org/" target="_blank" rel="noopener" title="orcid.org" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M294.75 188.19h-45.92V342h47.47c67.62 0 83.12-51.34 83.12-76.91 0-41.64-26.54-76.9-84.67-76.9M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8m-80.79 360.76h-29.84v-207.5h29.84zm-14.92-231.14a19.57 19.57 0 1 1 19.57-19.57 19.64 19.64 0 0 1-19.57 19.57M300 369h-81V161.26h80.6c76.73 0 110.44 54.83 110.44 103.85C410 318.39 368.38 369 300 369"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://github.com/cyverse-learning-material/" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://twitter.com/cyverseorg" target="_blank" rel="noopener" title="twitter.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://www.linkedin.com/company/cyverse.org" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3M135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5m282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://harbor.cyverse.org/" target="_blank" rel="noopener" title="harbor.cyverse.org" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1zm0-204.3h-66.1v60.7h66.1zm78.2 144.8H362v59.4h66.1zm-156.3-72.1h-66.1v60.1h66.1zm78.1 0h-66.1v60.1h66.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1zm78.1 0h-66.1v59.4h66.1zm-78.1-72.1h-66.1v60.1h66.1z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://www.youtube.com/c/CyverseOrgProject" target="_blank" rel="noopener" title="www.youtube.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 576 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M549.655 124.083c-6.281-23.65-24.787-42.276-48.284-48.597C458.781 64 288 64 288 64S117.22 64 74.629 75.486c-23.497 6.322-42.003 24.947-48.284 48.597-11.412 42.867-11.412 132.305-11.412 132.305s0 89.438 11.412 132.305c6.281 23.65 24.787 41.5 48.284 47.821C117.22 448 288 448 288 448s170.78 0 213.371-11.486c23.497-6.321 42.003-24.171 48.284-47.821 11.412-42.867 11.412-132.305 11.412-132.305s0-89.438-11.412-132.305m-317.51 213.508V175.185l142.739 81.205z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.indexes", "navigation.top"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.83f73b43.min.js"></script>
      
    
  </body>
</html>